feat: Add food price index generator and figure chart generator funct…

tonyshumlh opened the pull request:
@celestezhao please note that food price index generator function that you might ned
@JohnShiuMK @srfrew please note the figure chart generator function for frontend code and backend code design

tonyshumlh committed:
feat: Add food price index generator and figure chart generator functions

tonyshumlh committed:
fix: fix name in environment.yml

celestezhao commented:
@tonyshumlh 

Hi Tony, your code is awesome! 

Just one thing - when I use your generate_food_price_index_data() function, seems there is a mismatching between the desired `widget_date_range` and the actual date scope in `price_index_data` df:
<img width="783" alt="fdi_cal" src="https://github.com/UBC-MDS/DSCI-532_2024_19_food-price-tracker/assets/143828198/3c734830-09e8-4ac2-8819-a369a64005df">
Here I make the widget_date_range start from year 2010, and the starting year should be 2011 in returned price_index_data as our data starts from 2011. However, price_index_data returns 2014.

I'm thinking if this problem is due to `food_price_index_dict` - if the start date we expect does not have data for all the commodities defined in food_price_index_dict, then after executing the following code, records of these dates will be filtered out:
```
price_index_data = price_index_data[
    price_index_data.index_weight == len(food_price_index_dict)
].drop(columns=["index_weight"])
```
In the final selection of price_index_data, you require that records for each date must include all three commodities (Rice, Radish, Sugar). If a date is missing at least one of these commodities, then the record for that date will not be included in the final DataFrame.

Please lmk if you have some thoughts. Thanks again for calculating food price index, this has helped me a lot!

srfrew commented:
This code functions well given our testing dataset retrieved using the fetch_data functions in #49 

Will complete review shortly

srfrew commented on src/plotting.py:
This code and output is great! Thank you for such awesome work. 

Tested locally with fetch_data functions and functions as intended 
Unintended behaviour occurs if the date range is outside the provided data date range

srfrew reviewed (approved):
Code functions locally with fetch_data functions! Great work

srfrew commented:
Local testing during review: 

```python
from fetch_data import fetch_country_data, fetch_country_index
from plotting import generate_figure_chart, generate_food_price_index_data
import pandas as pd
import altair as alt
alt.data_transformers.enable('vegafusion')

df = fetch_country_data("Japan", country_index_df = fetch_country_index())
figure_chart = generate_figure_chart(df, ('2018-01-01', '2020-09-15'), ["Osaka", "Tokyo"], ["Rice", "Radish"])
figure_chart[0]
```

celestezhao committed:
feat: Build line chart feature (#48)

* feat: Add line chart for Food Price Index

Add function generate_line_chart_fpi()

* feat: Add line chart for selected commodities

Add function generate_line_chart_commodities()

srfrew committed:
Merge branch 'main' into build_figure_chart_feature

srfrew commented:
Merged YML conflict from #47. Code review complete, committing to main.

srfrew commented:
Again awesome work Tony!
